pandas/core/internals.py

```diff
@@ -3748,18 +3769,37 @@ def comp(s):
             for i, (s, d) in enumerate(zip(src_list, dest_list)):
                 new_rb = []
                 for b in rb:
-                    if b.dtype == np.object_:
-                        convert = i == src_len
-                        result = b.replace(s, d, inplace=inplace, regex=regex,
-                                           mgr=mgr, convert=convert)
-                        new_rb = _extend_blocks(result, new_rb)
+                    # regular expression support needs to be improved.
+                    # If the replacement for the previous pattern
+                    # matches the next pattern, the value will be replaced
+                    # again with a different value from dest_list
+                    # i.e. when values is ['a', 'b']
+                    # src_list: [r'a*', r'b*'], dest_list: ['b', 'a']
+                    # result will be ['b', b'] after searching for pattern r'a'
+                    # and then changed to ['a', 'a'] for pattern r'b*'
+                    if regex:
+                        if b.dtype == np.object_:
+                            convert = i == src_len
+                            result = b.replace(s, d, inplace=inplace,
+                                               regex=regex,
+                                               mgr=mgr, convert=convert)
+                            new_rb = _extend_blocks(result, new_rb)
+                        else:
```

Comment:
I kept the logic in master untouched for regex mode. This means regex replace still behave incorrectly (just like how dict replace is behaving now)
 ``_maybe_compare`` can only do equality compare for now. That is the cause of it. I haven't figure out how to add regex support in ``_maybe_compare`` without breaking any existing test. Once _maybe_compare is fixed, this part can be removed. I think @Licht-T is working on this part #20656.

Reply:
well what I would do is add a ``regex=`` param to ``_replace_coerce`` and push the logic to the block

Reply:
In addition to ``regex=`` param, it also needs ``src=`` param for the regex pattern to match. I think the ``_replace_coerce`` should only take a boolean array (``mask=``) indicating where to put the new value. Meanwhile, regex matching should be done at generating the mask before passing it to the ``_replace_coerce``.  Adding regex logic to it might makes it less clear for others and probably let someone else override ``_replace_coerce`` or reuse it for regex replace. What I want is keep ``_replace_coerce`` just like ``putmask`` and handle the regex comparison at ``_maybe_compare(values, s, operator.eq)``

Reply:
I disagree, this is crazy code here and needs simplification. all of the real work should be done in the blocks themselves, e .g. object is different than the other blocks. 

Here should be a simple
```
rb = [b._replace(.....) for b in rb]
```

you can pass whatever args you want, but the top level logic is just too complicated here